sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLDriver.scala

```diff
@@ -64,9 +65,15 @@ private[hive] class SparkSQLDriver(val context: SQLContext = SparkSQLEnv.sqlCont
         new VariableSubstitution().substitute(command)
       }
       context.sparkContext.setJobDescription(substitutorCommand)
-      val execution = context.sessionState.executePlan(context.sql(command).logicalPlan)
-      hiveResponse = SQLExecution.withNewExecutionId(execution) {
```

Comment:
Can we fix the problem inside `SQLExecution.withNewExecutionId`? So that the issue can be fixed in all the places, not only SQL shell. e.g. `sql("show tables").show` and also thrift-server.

Reply:
Ok, let me try.

Reply:
As discussed in https://github.com/apache/spark/pull/31968#discussion_r603155308
When run `sql("show tables").show`, we would see 3 SQL queries in Spark SQL UI:

Query0 `sql("show tables")`: execute the `ShowTablesCommand `

Query 1 `show`:  the logical plan is the same with query 0, triggered in `Dataset.getRows` https://github.com/apache/spark/blob/3b634f66c3e4a942178a1e322ae65ce82779625d/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L291

Query2 `show`:  show the result of the command in `LogicalRelation` 

We can check the `LogicalRelation.fromCommand` flag and don't produce `executionId` for query2 in `SQLExecution.withNewExecutionId`,  any thoughts?

Reply:
I was wondering if it's better to remove Query1 from the SQL UI, as Query2 is what `show` actually does.  But I didn't come up with a good solution.

Reply:
`toDF()` triggers a SQL execution? Do you know how that happens?

Reply:
`toDF` only produces a `executionId` by `SQLExecution.withNewExecutionId`, but the `command`  with side effects would run only once.  It happens as below:
1. when run `sql("show tables")`, Spark generates a `QueryExecution` object. The `QueryExecution`'s `LogicalPlan` is eagerly executed when we pass the  `QueryExecution` to a Dataset and initialize `Dataset.logicalPlan`
2. when run `show`,  the code goes to `Dataset.showString`https://github.com/apache/spark/blob/3b634f66c3e4a942178a1e322ae65ce82779625d/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L340
and then call `Dataset.getRows`
https://github.com/apache/spark/blob/3b634f66c3e4a942178a1e322ae65ce82779625d/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L291
we new a `Dataset` with the same `QueryExecution` object in _**step1**_
https://github.com/apache/spark/blob/3b634f66c3e4a942178a1e322ae65ce82779625d/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L454
thus, the `command`  would be wrapped in `SQLExecution.withNewExecutionId` when initializing the `df.logicalPlan`  , however,  the `QueryExecution` has already been optimized and planned, won't execute the underlying command any more, see `ExecutedCommandExec.sideEffectResult`
https://github.com/apache/spark/blob/3b634f66c3e4a942178a1e322ae65ce82779625d/sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala#L68

Reply:
how about we update `toDF()` and return `this` directly if the encoder is already a `RowEncoder`? We can check `Dataset.encoder.clsTag`

Reply:
And I think it's better if `sql("show tables").show` only triggers one SQL execution. Query 2 should also be removed from the UI.

Reply:
When run `sql("show tables").collect`, there are 2 same SQL queries in SQL UI: execute `ShowTablesCommand `. This is different from running "show tables" in spark-sql.

The reason is that:
1. run "show tables" in spark-sql
we create a DataFrame from the "show tables" SQL,  its `logicalPlan` is a `LocalRelation`. Then create a `QueryExecution` from the `LocalRelation` and convert it to Hive result string.
see`SaprkSQLDriver.run`
https://github.com/apache/spark/blob/06c09a79b371c5ac3e4ebad1118ed94b460f48d1/sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLDriver.scala#L67-L70
the query plan is to collect from `LocalRelation`

2. run `sql.("show tables").collect`
When call collect, the `Dataset` directly use its own `QueryExecution` , see https://github.com/apache/spark/blob/06c09a79b371c5ac3e4ebad1118ed94b460f48d1/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L2983
the query plan is just to execute the `ShowTablesCommand`

So it seems that we can't solve the problem by checking the `LocalRelation.fromCommand` in `SQLExecution. withNewExecutionId `

I'm thinking another way:
1. unify the behavior in `SaprkSQLDriver.run` and `Dataset.collect`
don't create extra `QueryExecution` in `SaprkSQLDriver.run` (what's the original purpose to create a new `QueryExecution`?)
``` 
val execution = context.sql(command).queryExecution
```
2. add a `isCommand` flag in `Dataset`
3. pass the `Dataset` to `SQLExecution. withNewExecutionId` and check `Dataset.isCommand`

WDYT?

Reply:
> don't create extra QueryExecution in SaprkSQLDriver.run

+1 to fix this.

One idea is to move the eager command execute to `QueryExecution`, so that `withNewExecutionId` can know if it's a command execution or not from `QueryExecution` instance.